File: 1-01 FEN counting C/out_GPT4.c
#include <stdio.h>
#include <string.h>

void countFEN(int *pieces, char *fen) {
    // Reset the pieces array to ensure it starts from zero
    for(int i = 0; i < 12; i++) {
        pieces[i] = 0;
    }
    
    // Iterate over the FEN string to count pieces
    for(int i = 0; fen[i] != ' ' && fen[i] != '\0'; i++) {
        switch(fen[i]) {
            case 'K': pieces[0]++; break;
            case 'Q': pieces[1]++; break;
            case 'R': pieces[2]++; break;
            case 'B': pieces[3]++; break;
            case 'N': pieces[4]++; break;
            case 'P': pieces[5]++; break;
            case 'k': pieces[6]++; break;
            case 'q': pieces[7]++; break;
            case 'r': pieces[8]++; break;
            case 'b': pieces[9]++; break;
            case 'n': pieces[10]++; break;
            case 'p': pieces[11]++; break;
            default:
                // Handle numbers (empty squares) by doing nothing
                break;
        }
    }
}

int main() {
    // Example usage
    int pieces[12];
    char fen[] = "rnbqkbnr/pppppppp/8/8/8/8/PPPPPPPP/RNBQKBNR w KQkq - 0 1";
    
    countFEN(pieces, fen);
    for(int i = 0; i < 12; i++) {
        printf("%d ", pieces[i]);
    }
    
    return 0;
}

File: 1-02 Fen Counting Py/out_GPT4.py
def countFEN(fen):
    # Define a dictionary to keep track of the counts for each piece.
    piece_counts = {
        'K': 0, 'Q': 0, 'R': 0, 'B': 0, 'N': 0, 'P': 0,  # White pieces
        'k': 0, 'q': 0, 'r': 0, 'b': 0, 'n': 0, 'p': 0  # Black pieces
    }
    
    # Iterate through the FEN string to count pieces
    for char in fen:
        if char in piece_counts:
            piece_counts[char] += 1
    
    # Arrange the counts in the specified order: King, Queen, Rook, Bishop, Knight, Pawn
    white_counts = [piece_counts['K'], piece_counts['Q'], piece_counts['R'], piece_counts['B'], piece_counts['N'], piece_counts['P']]
    black_counts = [piece_counts['k'], piece_counts['q'], piece_counts['r'], piece_counts['b'], piece_counts['n'], piece_counts['p']]
    
    # Concatenate white and black counts into one array and return it.
    return white_counts + black_counts

# Example usage:
fen = "rnbqkbnr/pppppppp/8/8/8/8/PPPPPPPP/RNBQKBNR w KQkq - 0 1"  # Starting position
print(countFEN(fen))
